File: LeetCode/Kotlin/s131.kt
class Solution131 {
    fun partition(s: String): List<List<String>> {
        val pars = mutableListOf<List<String>>()
        fun calc(start: Int, par: MutableList<String>) {
            val n = s.length
            if (start == n) pars.add(par.toList())
            else {
                (start..<n).forEach{ i ->
                    if (s.isPalindrome(start, i)) {
                        par.add(s.slice((start..i+1)))
                        calc(i+1, par)
                        par.removeLast()
                    }
                }
            }
        }
        calc(0, mutableListOf())
        return pars
    }

    private fun String.isPalindrome(l: Int, r: Int): Boolean {
        var l = l
        var r = r
        while (l<r) if (this[l++] != this[r--]) return false
        return true
    }

}

File: LeetCode/Rust/s131.rs
struct Solution131{}

impl Solution131 {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let mut pars: Vec<Vec<String>> = vec![];
        let mut par:Vec<String> = vec![];
        let s = s.as_bytes();
        Self::calc(s, 0, &mut par, &mut pars);
        return pars;
    }

    fn calc(s: &[u8], start: usize, par: &mut Vec<String>, pars: &mut Vec<Vec<String>>) -> () {
        let n = s.len();
        if start == n {pars.push(par.clone())}
        else {
            (start..n).for_each(|i|{
                if Self::is_palindrome(s.clone(), start, i) {
                    par.push(String::from_utf8(s[start..=i].to_owned()).unwrap());
                    Self::calc(s, i+1, par,  pars);
                    par.pop();
                }
            })
        }
    }

    fn is_palindrome(s: &[u8], mut l: usize, mut r: usize) -> bool {
        while l<r {
            if s[l] != s[r] {return false};
            l += 1;
            r -= 1;
        }
        return true
    }
}